### PR TITLE
test: Refactor tests

### PR DESCRIPTION
**Pull request by lynnagara**
This PR introduces BaseStorageTest to make it easier to test a storage
without an accompanying dataset (not used yet but intended to allow for
the deletion of the errors dataset). It also simplifies BaseTest and
ensures that migrations run before every test that inherits from this,
not just those for which a dataset name is passed in the setup_method.

### CODE REVIEW

File: tests/subscriptions/test_worker.py [hunk 1]
```diff
@@ -43,15 +46,8 @@ def all(self) -> Iterable[Tuple[UUID, SubscriptionData]]:
         return [*self.__subscriptions.items()]
 
 
-@pytest.fixture
-def dataset() -> Iterator[Dataset]:
-    with dataset_manager("events") as dataset:
-        yield dataset
```
Comment (fpacifici) [inline tests/subscriptions/test_worker.py:49]: Why removing the fixture? It is a convenient way to reuse the logic that prepares a dataset with its storages and tear it down across tests.

File: tests/base.py [hunk 2]
```diff
@@ -1,66 +1,69 @@
...
-        self.database = os.environ.get("CLICKHOUSE_DATABASE", "default")
-        self.dataset_name = dataset_name
 
-        if dataset_name is not None:
-            self.__dataset_manager = dataset_manager(dataset_name)
-            self.dataset = self.__dataset_manager.__enter__()
-        else:
-            self.__dataset_manager = None
-            self.dataset = None
+class BaseStorageTest(object):
```
Comment (fpacifici) [inline tests/base.py:49]: I am not sure I see a great value in preserving a class based organization for these tests.
It may have been useful earlier on, but it seems we could have a more effective reuse by actually moving away from classes and using fixtures instead.
Reasons:
- we should be able to set up and truncate a dataset (or a storage) in the same way for any test that uses them while now we have a mix of initialization done in stup/teardown methods in the classes and ad-hoc or fixture based initialization done in the module based tests. If we converge to one approach only, then I would argue we would go for fixtures and classes here would not be giving anything other than grouping tests together (which modules can do as well).
- By not relying on class inheritance to inherit the dataset initialization logic we would encourage people to write reusable fixtures that could be used in tests that do not extend from the same class hierarchy.

File: tests/base.py [hunk 3]
```diff
@@ -1,66 +1,69 @@
...
-        if self.__dataset_manager:
-            self.__dataset_manager.__exit__(None, None, None)
 
-        redis_client.flushdb()
+class BaseDatasetTest(object):
+    def setup_method(self, test_method, dataset_name: str) -> None:
+        _setup()
+        self.dataset_name = dataset_name
+        self.dataset = get_dataset(dataset_name)
+        from snuba.web.views import truncate_dataset
```
Comment (evanh) [inline tests/base.py:63]: Out of curiousity, why is this import here instead of up top?

File: tests/base.py (hunk 2 above)
Comment (tkaemming) [reply tests/base.py:49]: Strongly agree with @fpacifici's comment. One of the biggest issues I have with class-based tests is that it's very easy to accidentally bring in a lot of unnecessary preparation of fixture state from a superclass (how many things that subclass `BaseTest` actually need all this stuff?), having to explicitly reference fixtures in the test definition makes things a lot more efficient (and easier to understand in my opinion.)

**Comment by lynnagara** [comment]
Superceded by https://github.com/getsentry/snuba/pull/1382
